Fix inplace modification occuring in Arc.intersections

ghiggi opened the pull request:
This PR fixes a bug (inplace modification) occurring in `Arc.intersections` method. 

 - [ ] Closes #xxxx <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - https://github.com/pytroll/satpy/issues/383
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` <!-- remove if you did not edit any Python files  -->

gerritholl commented on pyresample/spherical.py:
```diff
@@ -253,12 +254,16 @@ def intersections(self, other_arc):
         From http://williams.best.vwh.net/intersect.htm
         """
         if self.end.lon - self.start.lon > np.pi:
+            self = copy.deepcopy(self)
```
I don't know if one should copy here, but calling a copy of self `self` may be confusing (I've never seen it before), maybe you can give a different name to the copy?  `new = copy.deepcopy(self)` maybe?

djhoese replied:
I tried to mention this in the issue you commented on @ghiggi, but I don't think you need `self` here at all. If you start out this method with gather the lon/lat values of each arc point then `self` and `other_arc` ~can remain untouched as they aren't passed to anything after this~. Ah I see they are used in cross2cart, but I'm not sure that matters as it is the start SCoordinate that is used

```python
# these are scalars, right?
end_lon = self.end.lon
other_end_lon = other_arc.end.lon

... if statements that modify end_lon and other_end_lon ...

ea_ = self.start.cross2cart(SCoordinate(end_lon, self.end.lat)).normalize()
eb_ = other_arc.start.cross2cart(SCoordinate(other_end_lon, other_arc.end.lat)).normalize()
```

djhoese reviewed (commented):
@mraspaud the `spherical_geometry.py` is completely deprecated now right? Are we using that anywhere? Can @ghiggi just remove it (probably in another PR)?

djhoese reviewed (changes requested):
Looks good. Could you add a test for the main `spherical.py` logic to make sure you are fixing what you think you are? Maybe based on the case described in the related issue?

Depending on what @mraspaud says, you can probably remove the updates to `spherical_geometry.py`.

ghiggi commented:
@djhoese spherical geometry is not deprecated. It is still used unfortunately by `BaseDefinition` (and inherited by `SwathDefinition` and `AreaDefinition`) `contains`, `corners`, `overlaps`,  `intersection`, `overlap_rate` and `get_area` methods !!!
 See https://github.com/pytroll/pyresample/blob/main/pyresample/geometry.py#L415

ghiggi commented:
I don't have the data to reproduce the satpy issue ....

djhoese commented:
Ok thanks for the info. For the issue data, that's why this would be a unit test. The goal would be to write a test with fake data that resembles the original use case (fake lon/lats that cross the anti-meridian). Prove that the test fails as expected before your changes here (comment them out locally) then add your changes and prove that they make the test pass.

ghiggi commented:
Sorry ... I am not used to such testing frameworks. Do you mean that the modification in place does not occur anymore, or related to the goes17 disk ...

ghiggi commented:
Something like this? 
```
from pyresample.spherical import Arc, SCoordinate 
arc1 = Arc(SCoordinate(*np.deg2rad((143.76, 0))), 
           SCoordinate(*np.deg2rad((143.95, 7.33)))
           )
arc2 = Arc(SCoordinate(*np.deg2rad((170.34, 71.36))), 
           SCoordinate(*np.deg2rad((-171.03, 76.75)))
           )
arc2_orig = copy.deepcopy(arc2)
_ = arc1.intersection(arc2)
assert arc2_orig.end.lon == arc2.end.lon 
```

djhoese commented:
Yes, but you could go one step further. In software testing we may want to make low-level assertions like you have done, but we also want to make sure the high-level issue is resolved. So on the low-level, making sure that the original object was not modified. That is definitely a side effect that should not happen, especially from a method called `intersection`. On the high-level it might be nice to have a test between two polygons as @pnuu had in #383 (the global EPSG:4326 and the geostationary satellite full disk) and make sure that the intersection is where/what we expect. A simpler version of that test (but also less complete) would be to make sure that `arc1.intersection(arc2)` returns the expected result in your test case. Right now you are throwing the result away.

So I think at a minimum I would like the assertion that the intersection returns an exact result and I would like (as you have it in your example) a check that the original arc is unaffected. If you can add the polygon test that would be great. Side question, do we know that `deepcopy` is able to properly copy an Arc object?

djhoese commented:
Yep, deepcopy looks like it works.

```python
In [1]: from pyresample.spherical import Arc, SCoordinate

In [2]: start = SCoordinate(0, 0)

In [3]: end = SCoordinate(1, 1)

In [4]: arc = Arc(start, end)

In [5]: import copy

In [6]: arc_orig = copy.deepcopy(arc)

In [7]: arc.start.lon = 100

In [8]: arc
Out[8]: (5729.5779513082325, 0.0) -> (57.29577951308232, 57.29577951308232)

In [9]: arc_orig
Out[9]: (0.0, 0.0) -> (57.29577951308232, 57.29577951308232)
```

mraspaud commented:
> @mraspaud the `spherical_geometry.py` is completely deprecated now right? Are we using that anywhere? Can @ghiggi just remove it (probably in another PR)?

I'm not sure. But there seems to be some duplicated code, can we refractor it in some common functions?

ghiggi commented:
I added the tests and fixed another issue related to testing SCoordinate equality at the antimeridian. I also had to change a test because was using "invalid" longitudes. 

I realized that we should keep open the [satpy issue 383]( https://github.com/pytroll/satpy/issues/383 ) because it's unclear to me what should be the expected intersection polygon between GOES-17 and EPSG_4326. In the new spherical API it can return 2 polygons, one for both sides of the antimeridian line.

djhoese commented on pyresample/test/test_spherical.py:
```diff
@@ -308,6 +325,41 @@ def test_angle(self):
+        # Assert disjoint arcs returns None
+        assert isinstance(point, type(None))
+
+        # Test intersecting arcs across the antimeridian
```
Can you split this second half into another test method? Or we could get fancy and I could show you how to do pytest parametrize stuff, but we'd have to put this test in another (non-unittest.TestCase) class.

ghiggi replied:
Now I split into two functions. But would be nice if we can meet once and you explain to me briefly the best practices so that I can use them for the new spherical tests ...

djhoese replied:
Yes, I could meet if needed. I'm not sure I have too much to teach, but if you're OK with asynchronous chatting we could talk about this in the pytroll slack's `#code` channel.

As for this suggestion, it may not be obvious why I asked to do this compared to other tests in pyresample. A lot of the existing tests in pyresample are following bad practice. Additionally a lot of the tests are written with the standard library's `unittest` which is not as flexible compared to pytest which we've been using more and more throughout the pytroll libraries. In general, we want to keep clean and short test methods so they are easy to understand and easy to modify if needed. This should result in less duplicated code if the tests all depend on the same helper functions (for example, to create fake data). It also helps in seeing at a glance what is failing and why if the test has a single purpose. For example, if I run the pyresample tests and your original function fails, I don't know (initially) if the first chunk failed or the second. And even if I did I have to read the source code and understand the entire test function to understand how the various sections differ. The name of the test function/method is usually shown when there is a failure so if I see that and it is descriptive I can usually get a basic idea of what is going on and why the tests are failing.

ghiggi replied:
OK with asynchronous chatting we could talk about this in the pytroll slack's #code channel. 
Just create/add me to that channel ;)

djhoese replied:
I think you can join whatever channels you want on the slack :wink:

djhoese commented on pyresample/test/test_spherical.py:
```diff
@@ -605,6 +656,74 @@ def test_intersection(self):
+                                                  [0, 90],
+                                                  [180, 90],
+                                                  [180, -90]]))
+        vertices_goes_west = np.array([[2.50919361e+00, 1.19782309e-16],
```
Do we need *all* these points for this test to work and make sense? Meaning, does it fail before your changes if it only had 5 or 6 points?

ghiggi replied:
Do you mean if I can simply the goes-west vertices @djhoese ?

djhoese replied:
Yes. Can it use less vertices?

ghiggi replied:
Done

djhoese reviewed (changes requested):
Looking really good. I've made a couple requests in the comments below that I think would help. Otherwise, I have two big picture questions:

1. You added the radian wrapping in the SCoordinate equality. This is going to slow down spherical operations even more. You also updated the one test because it had "invalid" longitudes. The longitudes are expected to be between -180 and 180, right? And the final results of calculations are always in -180 to 180? Would it be possible to remove the unwrap_radians stuff if we assume all inputs are -180 to 180? Alternatively, could we assume the SCoordinate always gets valid longitudes and do unwrapping at a higher level like at the polygon? I realize this would make SCoordinate allow for invalid coordinates, but to me that's a necessary evil at the moment if it means faster code.
2. Based on the Satpy issue and what Panu was trying to do, would it be possible (even if only in the new API) to return a "merged" version of the multiple intersections...no that's probably a bad idea. I'm OK leaving that issue open if it is still unclear, but a comment in the near future would be nice to clarify what is confusing and where the future of the issue is going.

ghiggi commented:
Answering your questions @djhoese. 

1. I added the radian wrapping in the SCoordinate equality because is required to identify that a point (-180,75) is equal to (180,75) !!! Otherwise, this case can raise issues in the arc/polygon intersection methods (i.e. when an intersecting point is located at -180 or 180 in longitude !). It's true that simultaneously it also bounds the longitudes to -180/180 ... even better ... 
 
2. In the new API would be very easy to try performing a union on the resulting intersection geometries. We will rediscuss it in due time when I do the PR on the SPolygon/SMultiPolygon ;)

mraspaud commented on pyresample/spherical_geometry.py:
```diff
@@ -236,17 +236,23 @@ def angle(self, other_arc, snap=True):
 
     def intersections(self, other_arc):
         """Get the two intersections of the greats circles defined by the current arc and *other_arc*."""
+        end_lon = self.end.lon
```
This function and the one in spherical.py are really similar, could the common code be factorized?

ghiggi replied:
A sort of. Coordinate can accept more types of args (in an unsafe way): 
The problem is rather that `spherical_geometry.Arc` methods seem to not work really well. 
There is also a test failing ...

mraspaud replied:
Ok, I see now that we have marked spherical_geometry.py for deprecation in pyresample 2.0. I suggest we leave that module as it is, and focus on spherical.py only.

djhoese replied:
As far as deprecation, @ghiggi has pointed out in other threads that it is still heavily used in the geometry classes. We'll need to work on that.

djhoese replied:
I guess I was hoping the updates to the geometry classes would be simple like changing the import and the class name used.

djhoese commented:
> I added the radian wrapping in the SCoordinate equality because is required to identify that a point (-180,75) is equal to (180,75) !!! Otherwise, the old behavior can cause errors in the arc/polygon intersection methods (i.e. when an intersecting point is located at -180 or 180 in longitude !). It's true that simultaneously it also bounds the longitudes to -180/180 ... even better ...

:thinking: But does the other code in the spherical code generate longitudes outside this range? If we document and assume that inputs are between -180 and 180 (wherever the closed or open side of the interval has to be, then could this unwrapping be removed? Put another way, if we decide not to support invalid longitudes, is there anything else in the logic of this code that would need this unwrapping? Otherwise, could this wrapping/unwrapping be done on the SCoordinate `__init__` instead? Or as I suggested before, only be done on the polygon class?

ghiggi commented:
> Does the other code in the spherical code generate longitudes outside this range? 

Now that the in-place modification does not occur anymore, I think no values are generated outside the `]-180,180[` range.

> If we document and assume that inputs are between -180 and 180 (wherever the closed or open side of the interval has to be, then could this unwrapping be removed? Put another way, if we decide not to support invalid longitudes, is there anything else in the logic of this code that would need this unwrapping? 

If inputs are between `[-180, 180[ ` the answer is yes, it can be removed.
If the inputs are between `[-180, 180]`  the unwrapping can be removed! The "unwrapping" is needed to be sure that the points with the valid longitudes (180, lat) and (-180, lat) are considered the same point. 
 
> Otherwise, could this wrapping/unwrapping be done on the SCoordinate __init__ instead? 

This is actually a very good point. I moved the unwrapping to the `SCoordinate.__init__ `.

> Could this wrapping/unwrapping be done only be done on the polygon class?

Could not be done only in `SphPolygon`. But I now added it also to `SphPolygon.__init `.

djhoese commented:
> If inputs are between [-180, 180[ the answer is yes, it can be removed.
If the inputs are between [-180, 180] the unwrapping can be removed! 

Ok so if we don't accept invalid longitudes, then all of this unwrapping can be removed? You're saying that the code as it was before considered -180 and 180 the same coordinate? But then you say:

> The "unwrapping" is needed to be sure that the points with the valid longitudes (180, lat) and (-180, lat) are considered the same point.

So it is needed? Or is it only needed for this -180/180 case and works without the unwrap just fine for other invalid longitudes?

ghiggi commented:
I was not clear sorry. 
> Does the code as it was before considered -180 and 180 the same coordinate? 

No. This was a possible source of errors and it has been solved.

> If we don't accept invalid longitudes, then all of this unwrapping can be removed? 

I am not sure of the answer. For "accepting" do you mean raising an error if not within valid longitudes or unwrapping in the `__init__`? Because methods such `np.arctan2` in `Arc.intersections` can returns values between [-180, 180], so some unwrapping is still needed to ensure [-180, 180[. 

> So it is needed? Or is it only needed for this -180/180 case and works without the unwrap just fine for other invalid longitudes?

Is for sure needed for the -180/180 case. 
Values outside [-180, 180] are not generated anymore by the methods.

djhoese commented:
> For "accepting" do you mean raising an error when __init__ the geometry or unwrapping in the __init__?

I mean adding to the docstring "Vertex coordinates are assumed to be in the range [-180, 180[" and then assume that this is true of the inputs in the code. Removes complexity and improves performance and the cost of flexibility for what input will work. I ended up doing something similar for the SwathDefinition years ago because we were always checking if the lon/lats were in the -180/180 range, but this is a huge performance hit if it isn't needed *especially* large swaths and especially on `__init__` if the SwathDefinition coordinates are never used.

So all of that said, it sounds like we need to handle the -180 == 180 case no matter what as it is very likely (in my opinion) that people will give coordinates with both of these.

djhoese commented:
Looks like your tests are failing (see failing CI "Details" above). Also could you change usage of `self.assertTrue` and `self.assertFalse` to `assert X` and `assert not X` in your new tests? It will mean fewer changes if/when we switch to pytest.

ghiggi commented:
Ok now also the `test_spherical_geometry.py` pass ;)  
`Coordinate` expects lon/lat in degree, but then `self.lon/self.lat` returns radians ... quite weird!

I did modify the test to migrate toward pytest. 
CodeFactor seems to not be happy with the following:

 ```
 def test_eq(self):
        """Test equality of two cartesian vectors."""
        d = CCoordinate((1., 0., 0.))
        c = CCoordinate((0., 1., 0.))
        assert not (c == d)
```
Should I modify in `assert (c != d)`? But then we are testing the `__neq__` method instead of `__eq__`. Let me know

ghiggi commented:
@djhoese ready to go I think ;)

djhoese commented on pyresample/test/test_spherical.py:
```diff
@@ -106,6 +106,17 @@
+        from pyresample.spherical import modpi, unwrap_radians
+
+        # Test points at np.pi are converted to -np.pi
+        assert np.allclose(modpi((-np.pi, np.pi)), (-np.pi, -np.pi))
+        assert np.allclose(unwrap_radians((-np.pi, np.pi)), (-np.pi, -np.pi))
```
```suggestion
        np.testing.assert_allclose(modpi((-np.pi, np.pi)), (-np.pi, -np.pi))
        np.testing.assert_allclose(unwrap_radians((-np.pi, np.pi)), (-np.pi, -np.pi))
```

djhoese commented on pyresample/test/test_spherical.py:
```diff
@@ -117,7 +128,7 @@ def test_distance(self):
         """Test Haversine formula."""
         d = SCoordinate(0, 0).hdistance(SCoordinate(1, 1))
-        self.assertTrue(np.allclose(d, 1.2745557823062943))
+        assert (np.allclose(d, 1.2745557823062943))
```
```suggestion
        np.testing.assert_allclose(d, 1.2745557823062943)
```

djhoese replied:
To save me time...could you do this same update to all your use cases in the tests and remove the unnecessary surrounding `()` in your assert statements?

djhoese commented on pyresample/spherical.py:
```diff
@@ -26,6 +26,19 @@
+
+def unwrap_radians(val, mod=np.pi):
+    """Put *val* between -*mod* and *mod*."""
+    val = np.asarray(val)
```
Do we need this here? What are the cases where this isn't a numpy array?

ghiggi replied:
In `spherical.py`, `val` is either a float or a `np.array `. But if I provide a tuple of longitudes as input, then it would fail. Should I remove it @djhoese ?

djhoese replied:
Is there code in `spherical.py` that passes a tuple? If your code only ever passes a scalar float or a numpy array then I'd say remove this. You could go even further and change the name of this to `_unwrap_radians` to identify to the reader of the code (in the future) that this function shouldn't be used outside of this module.

djhoese commented on pyresample/test/test_spherical.py:
```diff
@@ -383,41 +429,41 @@ def test_is_inside(self):
         polygon2 = SphPolygon(np.deg2rad(vertices))
 
-        self.assertTrue(polygon1._is_inside(polygon2))
+        assert (polygon1._is_inside(polygon2))
```
These parentheses aren't necessary. Could you please remove them on this line and all other assert lines?

djhoese commented:
Do you have any comparisons of performance before this PR and with it? I normally wouldn't ask for a performance comparison in a PR for a bug fix, but you've added operations on a very low-level class that is created and used a lot in a lot of tight for loops.

ghiggi commented:
I quickly timed just some polygon intersections but I didn't notice any change. 
Here below an example between GOES-WEST (1km) and GOES-ABI (1km):
```
# 920 ms ± 35.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # before fix
# 870 ms ± 9.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # after fix
```
Should I do something more structured? On which methods? Do you have a template to follow for such stuffs @djhoese ?

djhoese commented:
That test is fine. I'm surprised it is faster, but also the fact that it takes almost a full second for this operation just shows how much this algorithm shouldn't be in pure python. Anyway...just the `tuple` question left.

djhoese commented on pyresample/spherical.py:
```diff
@@ -26,6 +26,18 @@
+EPSILON = 0.0000001
+
+
+def _unwrap_radians(val, mod=np.pi):
+    """Put *val* between -*mod* and *mod*."""
+    return (val + mod) % (2 * mod) - mod
+
+
+def modpi(val, mod=np.pi):
+    """Put *val* between -*mod* and *mod*."""
+    return _unwrap_radians(val, mod=mod)
```
Hopefully the last request: These functions are now exactly the same, right? If you think "unwrap radians" is a better name than "modpi" then delete "modpi". If not, delete `_unwrap_radians`. Replace whatever usage was deleted.

ghiggi replied:
@djhoese : `modpi` existed already so we need to keep it for backcompatibility. I think unwrap_radians is a better name ... what `modpi` stands for?

djhoese replied:
`modpi` shouldn't be public API so no user should be using it (I know it doesn't have a `_` prefix but it isn't intended for public use). It stands for modulo based on pi (I assume).

Put another way, this is a private function inside this module and should *only* be used by this module. If it is used somewhere else then we should probably fix that.

ghiggi replied:
@djhoese done

djhoese reviewed (approved):
Looks good to me. Nice job!

I will admit I was a little tougher on you than I might normally be with a newer contributor on a "small" PR like this, but given the number of changes you want to make elsewhere I thought it was good to get my preferences/opinions/lessons out right away. Thanks for sticking with it.

Once the tests pass I'll merge this.
